src/polyfills/intersection-observer.js

```diff
@@ -0,0 +1,276 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview
+ * See https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver.
+ */
+
+const UPGRADERS = '_upgraders';
+
+/**
+ * @param {!Window} win
+ */
+export function install(win) {
+  // DO NOT SUBMIT: remove, testing.
+  delete win.IntersectionObserver;
+  delete win.IntersectionObserverEntry;
+
+  if (!win.IntersectionObserver) {
+    win.IntersectionObserver = /** @type {typeof IntersectionObserver} */ (IntersectionObserverStub);
+  }
+
+  /* DO NOT SUBMIT: examples */
+  (() => {
+    const inob1 = new win.IntersectionObserver((records) => {
+      console.log('inob1: ', records);
+    });
+    inob1.observe(win.document.body);
+
+    const iframe = win.document.createElement('iframe');
+    iframe.srcdoc = '<!doctype html><html><head><body>';
+    iframe.onload = () => {
+      const childWin = iframe.contentWindow;
+      installForChildWin(win, childWin);
+      const inob2 = new childWin.IntersectionObserver((records) => {
+        console.log('inob2: ', records);
+      });
+      inob2.observe(childWin.document.body);
+
+      IntersectionObserverStub[UPGRADERS].push(() => {
+        setTimeout(() => {
+          const inob3 = new IntersectionObserverStub((records) => {
+            console.log('inob3: ', records);
+          });
+          inob3.observe(childWin.document.body);
+        }, 1000);
+      });
+    };
+    win.document.body.appendChild(iframe);
+  })();
+}
+
+/**
+ * @param {!Window} parentWin
+ * @param {!Window} childWin
+ */
+export function installForChildWin(parentWin, childWin) {
+  // DO NOT SUBMIT: remove, testing.
+  delete childWin.IntersectionObserver;
+  delete childWin.IntersectionObserverEntry;
+
+  if (!childWin.IntersectionObserver && parentWin.IntersectionObserver) {
+    childWin.IntersectionObserver = parentWin.IntersectionObserver;
+    childWin.IntersectionObserverEntry = parentWin.IntersectionObserverEntry;
+    if (parentWin.IntersectionObserver === IntersectionObserverStub) {
+      // Wait until new polyfill is available and override with the upgraded
+      // classes.
+      IntersectionObserverStub[UPGRADERS].push(() => {
+        childWin.IntersectionObserver = parentWin.IntersectionObserver;
```

Comment:
Curious if the InOb within a friendly iframe work out of the box?

Reply:
It does now. I patched it in https://github.com/w3c/IntersectionObserver/pull/419. And the cross-origin iframe support is coming with https://github.com/w3c/IntersectionObserver/pull/421